### ISSUE TITLE
Add function to decrypt via the sops tool

### ISSUE THREAD

**Issue by smeierhofer** [roboll/helmfile#929]
I would like to read a file that has been encrypted via sops, decrypt it, and base64 encode it so that it is in the format needed by the Helm Secrets template.

The reason to do this is that my application has several application configuration files and some of the configuration settings are sensitive and should be encrypted when saved on disk. Also, I have a PKCS 12 keystore that is a binary format and would also like encrypted on disk. So then I could use code such as `{{- readFile "keystore.p12.enc" | decrypt | b64enc | nindent 4 }}`.

Sops works well to decrypt as it integrates into Azure Key Vaults (and similar tech from other vendors) where the encryption keys are stored in the Azure key vault and only accessible to those who have been granted access.

I am not sure of the options to the sops command that would need to be exposed to the a "decrypt" function in helmfile.

**Comment by mumoshu** [comment]
@smeierhofer Hey! Thanks for the request.

I've read this twice and, honestly saying, I'm still unsure why you'd need this.

First of all, shouldn't your chart `b64enc` it rather than you do it in Helmfile? I believe that's the pattern [seen in many charts](https://github.com/helm/charts/search?q=b64enc&unscoped_q=b64enc).

If you can change your chart as such, I believe you're all set. Helmfile already have an out-of-box support for passing any value extracted from a SOPS-encrypted file. See #906, and for usage with SOPS specifically please read https://github.com/variantdev/vals#sops.

**Comment by smeierhofer** [comment]
I was thinking that I need to helmfile to b64enc the file because the file is binary. Specifically, the helmfile values gotmpl reads and decrypts the binary file, then b64 encodes it and passes the base-64 encoded data to the helm chart via a chart value. The helm chart includes the encoded data in a Kubernetes Secret, which requires the data be base64 encoded, which it already is.

The problem I'm trying to solve is where helmfile reads an entire folder of configurations files and passes those to the chart. In one case, these files are plain text Java .properties file and I use the following to do this:
```yaml
app:
  config:
    {{- range $file := ( exec "sh" (list "-c" "echo config/*")) | trim | splitList " " }}
      {{- $contents := readFile $file }}
      {{- $config := printf "%s: |\n%s" (base $file) (indent 2 $contents) }}
          {{- $config | nindent 4 }}
    {{- end }}
```
But some of these configuration files contain sensitive data and so they need to be encrypted on disk and so I'd like to use code such as this `{{- $contents := readFile $file | decrypt }}` in the above code to read the encrypted configuration file, decrypt it using sops, and include the results in the values passed to the helm chart. In lieu of a "decrypt" function, I wonder if something like this would work for my purposes, but I haven't tried it: `{{- $contents := (exec "sops" list ( "--decrypt" $file ) ) }}`.

**Comment by mumoshu** [comment]
> the helmfile values gotmpl reads and decrypts the binary file, then b64 encodes it and passes the base-64 encoded data to the helm chart via a chart value. The helm chart includes the encoded data in a Kubernetes Secret, which requires the data be base64 encoded, which it already is.

@smeierhofer Thanks! This makes sense, but...

> But some of these configuration files contain sensitive data and so they need to be encrypted on disk and so I'd like to use code such as this {{- $contents := readFile $file | decrypt }} in the above code to read the encrypted configuration file, decrypt it using sops, and include the results in the values passed to the helm chart

So do you, in the end, get the decrypted(cleartext) .properties files, which you want pass to the chart? For that case I can still suggest you generating helmfile.yaml like this:

```
  values:
  - app:
      config:
        file1: ref+sops://path/to/file1.enc
        file2: ref+sops://path/to/file2.enc
        # ...
```

**Comment by mumoshu** [comment]
In case you do encrypt a binary file(other than .properties) with sops, I think we can enhance the `ref+` syntax to support returning b64-encoded result:

```
  values:
  - app:
      config:
        binaryfile1: ref+sops://path/to/binaryfile1.enc?encode=base64
        file2: ref+sops://path/to/file2.enc
        # ...
```

**Comment by smeierhofer** [comment]
Yes, in the end I do get the cleartext .properties file, which I pass to the chart. I will try out the suggestions. I did not know that I could simply list references to files in the values section of the helmfile.yaml. Also, I like the suggested way to handle base64 encoding the results.

**Comment by mumoshu** [comment]
> I did not know that I could simply list references to files in the values section of the helmfile.yaml

@smeierhofer It's the latest major new feature to Helmfile so we lack doc :) Sorry about that. Please feel free to ask anything about it.

> Also, I like the suggested way to handle base64 encoding the results.

Great! But to be clear, you won't need it for now if https://github.com/roboll/helmfile/issues/929#issuecomment-549108576 works so that you can just defer b64enc to the chart template?

**Comment by smeierhofer** [comment]
What happens if I use `file: ref+sops://path/to/binaryfile1.enc` to reference a binary file? I wouldn't expect Helm to allow binary data in a helm chart value. But I have never tried it.

**Comment by mumoshu** [comment]
@smeierhofer It decrypts `path/to/binaryfile1.enc` with sops. If the original file was a .properties file, it's basically the same as `file: <the content of the original .properties file, which is a text file containing key=value pairs>`.
